WIP: Protect against brute force (frequent) login attempts

marefr opened the pull request:
This aims to solve issue #7616 

This introduces a new concept of login attempts. The idea is that we keep a log of invalid login attempts. We use this log when a user in question tries to authenticate to validate if it has a number X (currently 5) of invalid login attempts during the last Y time (currently 5 minutes). 
* If so, we temporary don't allow the user to login. 
* If not and the user in question uses invalid credentials, we add a row to the login attempt log.

There's also a cleanup of old login attempts step added to the cleanup job. Right now this will delete login attempts that are older than 10 minutes.

Right now, login attempts are not captured if ldap is enabled. The reason for not having this if ldap is enabled is that it feels like Grafana instances that has ldap enabled not are exposed to the public internet, but I may be wrong. I would really need some feedback on this.

Questions:
* Do we want/need possibility to configure certain aspects of this feature?
  * Enable/dísable the feature?
  * X, where X is the number of invalid attempts before temporarily blocked?
  * Y, where Y is the historic time stamp we use for counting X

marefr committed:
db: add login attempt migrations

marefr committed:
db: add possibility to create login attempts

marefr committed:
db: add possibility to retrieve login attempt count per username

marefr committed:
auth: validation and update of login attempts for invalid credentials

If login attempt count for user authenticating is 5 or more the last 5 minutes
we temporarily block the user access to login

marefr committed:
db: add possibility to delete expired login attempts

marefr committed:
cleanup: Delete login attempts older than 10 minutes

The cleanup job are running continuously and triggering each 10 minute

daniellee commented on pkg/login/auth.go:
```diff
@@ -2,22 +2,26 @@ package login
 var (
-	ErrInvalidCredentials = errors.New("Invalid Username or Password")
+	ErrInvalidCredentials   = errors.New("Invalid Username or Password")
+	ErrTooManyLoginAttempts = errors.New("Too many consequent incorrect login attempts for user. Login for user temporarily blocked.")
```
Change consequent to consecutive.

daniellee commented on pkg/login/auth.go:
```diff
@@ -64,3 +80,20 @@ func loginUsingGrafanaDB(query *LoginUserQuery) error {
+		return err
+	}
+
+	if loginAttemptCountQuery.Result >= 5 {
```
As discussed, this should probably be configurable.

daniellee commented on pkg/services/sqlstore/login_attempt.go:
```diff
@@ -0,0 +1,91 @@
+			return err
+		}
+
+		maxId = toInt64(result[0]["id"])
```
To skip this conversion code, you could try a different xorm method than sess.Query. You could try using Find or Get together with a Where clause.

bergquist commented:
> Right now, login attempts are not captured if ldap is enabled. The reason for not having this if ldap is enabled is that it feels like Grafana instances that has ldap enabled not are exposed to the public internet, but I may be wrong. I would really need some feedback on this.

I do remember someone reporting having all their accounts betting blocked by the LDAP server when exposed to a brute force attack. I have no strong opinions about enable/disable it for public installs but I think some might use it. Not sure we should support it. 

I'm not sure we should make the number of attempts, the time window or how long they are stored configurable. I think we can find sensible defaults instead. Convention over configuration/Less is more etc :)

How about using Exponential back off instead? ex https://support.citrix.com/article/CTX136565
We could allow up to 3 failed attempts the last 1h and then start enforcing a delay until the same IP can try again.

marefr committed:
fix typo: rename consequent to consequent

marefr commented:
> I do remember someone reporting having all their accounts betting blocked by the LDAP server when exposed to a brute force attack. I have no strong opinions about enable/disable it for public installs but I think some might use it. Not sure we should support it.

I think you may refer to this issue: #9900 regarding ldap and brute force attack.

> I'm not sure we should make the number of attempts, the time window or how long they are stored configurable. I think we can find sensible defaults instead. Convention over configuration/Less is more etc :)

Yeah I can definitely see the point of sensible defaults :+1:. Still good to be able to disable the feature though if there should be any problem for an example.

> How about using Exponential back off instead? ex https://support.citrix.com/article/CTX136565
We could allow up to 3 failed attempts the last 1h and then start enforcing a delay until the same IP can try again.

Interesting. Are you thinking of implementing this per username and ip address or only ip address?

bergquist commented:
> Yeah I can definitely see the point of sensible defaults :+1:. Still good to be able to disable the feature though if there should be any problem for an example.

:100: 

> Interesting. Are you thinking of implementing this per username and ip address or only ip address?

Excellent question. I guess both?

torkelo commented:
I think this looks very close to mergable. 

- Think we can easily make this apply to LDAP logins as well (just record failed attempts for ldap as well)

Question is what should be configurable and what should the defaults be.

Two options with defaults like?
- login_attempts_per_minute = 3
- login_cool_down = 10m

The look is (currently if I read the code correctly) based on username + IP. Think this is fine for an initial version. 

Thoughts?

bergquist commented:
I think that sounds like a good start. If people need better protection we can improve this later.

marefr commented:
> * Think we can easily make this apply to LDAP logins as well (just record failed attempts for ldap as well)

Agree

> The look is (currently if I read the code correctly) based on username + IP. Think this is fine for an initial version.

That's not quite correct. The IP address are currently only stored and not used in login attempt count, only username is. 

Didn't quite see the win of using both username and IP address for locking accounts. Issues regarding IP addresses, i.e. possibility to block certain IP addresses/subnets from accessing Grafana feels like something that you usually use a firewall for. 

Having the IP addresses stored in the login_attempts table can however be helpful to manually verify if some specific IP address are trying to brute force login and then block this in firewall. 

> Two options with defaults like?
> * login_attempts_per_minute = 3
> * login_cool_down = 10m

What I tried to describe above was that we have X and Y in the current implementation. 
* X is the number of invalid attempts before temporarily blocked
* Y is the historic time stamp we use for counting X

So what you refer to as *login_attempts_per_minute* is basically what X and Y is. To support *login_cool_down* I think we need to store the actual login attempts as a counter instead, together with a timestamp. This will allow us to have an account locked for X minutes since the last invalid login attempt that made the counter go above the *max allowed invalid attempts*.

As a comparison, the current implementation works like a sliding window. Example:
* 13:00:00 - invalid login attempt
* 13:01:00 - invalid login attempt
* 13:02:00 - invalid login attempt
* 13:03:00 - invalid login attempt
* 13:04:00 - invalid login attempt
* 13:04:05 - login attempt blocked (counter = 5, account locked)
* 13:05:01 - invalid login attempt (counter = 4)
* 13:05:02 - login attempt blocked (counter = 5, account locked)

marefr commented:
After discussion with @torkelo:

* We add a possibility to to disable this feature by configuration. It will be enabled per default.
* Make this feature apply to LDAP logins as well.
* Skip checking IP address for now, only store it.
* Some documentation, at least for the configuration parameter.

bergquist commented:
Sounds great! 

One minor detail. I think it's better to have the default value of a setting be false since its the default value for bools in golang. 

So instead of having `enabled = true` as default. use `disabled = false`
This decreases the chances for bugs due to invalid default values of settings. 

@torkelo @daniellee opinions? 

This is not a rule that we follow in Grafana but I think we should start doing it.

marefr commented:
Are you meaning that this feature should be disabled per default in configuration or are you referring to the naming of variables, i.e. feature_enabled vs feature_disabled?

bergquist commented:
I'm referring to the variable name. so `feature_disable = false` should be the default for the setting.

marefr committed:
auth: enable login attempt validation for ldap logins

eripe970 commented on pkg/login/auth.go:
```diff
@@ -64,3 +82,20 @@ func loginUsingGrafanaDB(query *LoginUserQuery) error {
+func validateLoginAttemptCount(username string) error {
+	loginAttemptCountQuery := m.GetUserLoginAttemptCountQuery{
+		Username: username,
+		Since:    time.Now().Add(time.Minute * -5),
```
Should it be possible to configure this as well?

marefr replied:
No, not in the initial implementation

marefr committed:
auth: disable login attempts validation by configuration

Setting is named DisableLoginAttemptsValidation and is false by default
Config disable_login_attempts_validation is placed under security section
#7616

marefr committed:
auth: don't run cleanup of login attempts if feature is disabled

#7616

marefr committed:
auth: rename settings.go to ldap_settings.go

marefr committed:
auth: refactor AuthenticateUser

Extract grafana login, ldap login and login attemp validation together
with their tests to separate files.
Enables testing of many more aspects when authenticating a user.
#7616

marefr commented on conf/defaults.ini:
```diff
@@ -174,6 +174,9 @@ disable_gravatar = false
 data_source_proxy_whitelist =
 
+# disable protection against brute force (frequent) login attempts
+disable_login_attempts_validation = false
```
I don't know if this a good name

torkelo replied:
`disable_brute_force_login_protection` seems more clear

marefr replied:
Done. Thanks

marefr commented on pkg/setting/setting.go:
```diff
@@ -75,13 +75,14 @@ var (
+	DisableGravatar                bool
+	EmailCodeValidMinutes          int
+	DataProxyWhiteList             map[string]bool
+	DisableLoginAttemptsValidation bool
```
I don't know if this a good name

marefr reviewed (commented):
A bit unsure of the naming of configuration and setting variable

marefr committed:
auth: rename login attempt validation to brute force login protection

Setting DisableLoginAttemptsValidation => DisableBruteForceLoginProtection
Configuration disable_login_attempts_validation => disable_brute_force_login_protection
#7616

eripe970 commented on pkg/api/login.go:
```diff
@@ -102,12 +102,13 @@ func LoginPost(c *middleware.Context, cmd dtos.LoginCommand) Response {
-		Password: cmd.Password,
+		Username:  cmd.User,
+		Password:  cmd.Password,
+		IpAddress: c.Req.RemoteAddr,
```
Will this pick up up the `X-Forwarded-For` header if available?

torkelo replied:
it should (I hope so at least)

marefr replied:
According to macaron source code it does: https://github.com/Unknwon/macaron/blob/master/context.go#L133

faxm0dem commented:
This is still missing from the online documentation

vlada-jerkovic commented:
What are you thinking about implementing 3 step Login? This approach has Amazon and works pretty well.
1. Login page with only **username field and Submit button** -> redirect to
2. Page with Security check with **image, input field and Submit button** -> redirect to
3. Page with **Password field and Submit button**
Example:
![image](https://user-images.githubusercontent.com/20494777/115474365-94e9ff00-a23d-11eb-89e9-23315ca4a095.png)
